Tesseract Documentation and Tutorials

steviedale opened the pull request:
Updating Tesseract documentation and tutorials

mpowelson commented:
I haven't looked at this yet, but I am very excited about it. Thanks for doing this! Is this already building anywhere or do we need to build it locally to view it until we merge?

steviedale commented:
Hey @mpowelson! Glad to hear you're excited about it. Kevin and I are working with Levi on it and are hoping to get a good chunk of it done by mid next week. It isn't built anywhere yet, but it is setup to build locally.

mpowelson commented:
I think we need to add recommonmark to the requirements.txt. 

Here are the instructions I followed to build it

1.` pip3 install -r gh_pages/requirements.txt`
2. `sphinx-build gh_pages built_docs`

steviedale commented:
@mpowelson good catch! I just added it in 3a6f917

mpowelson commented on gh_pages/_source/tesseract_motion_planning_api_doc.rst:
```diff
@@ -0,0 +1,168 @@
+Tesseract Motion Planning API
+*****************************
+
+Overview
```
I think I might consider restructuring this. Rather than just stepping through a tesseract_ros example, it might be more useful to describe the pieces of a motion planning system and how we will interact with them.

I think I'd give an overview of the pieces like this with links to more details

- Environment Monitor - Description and link to page for modifying the environment
- Planning Server - Description
- Taskflows - What are available out of the box, and how do you define your own? Probably its own page too
- Command language - How to define a simple program. Link to more detailed docs

Then a simple example (e.g. <100 lines) that shows how to load a urdf, setup the planning server, and send it a request.The freespace_manager_example in tesseract_process_managers would probably be a good one to go off of.

steviedale replied:
I agree, I think that would be more helpful to have each component described and to use a simpler example that doesn't have other functionality distracting from the motion planning concept.

I rewrote this page in 81e5c20. It still needs some work, but it's an improvement from the last version. I used the example you pointed out as well. For now, I've omitted the description of taskflow until I have a better understanding of how it works and can describe it sufficiently.

Let me know what you think of this version.

marip8 commented on gh_pages/_source/getting_started_doc.rst:
```diff
@@ -0,0 +1,67 @@
+
+  cd ~/tesseract_ws/src 
+
+Clone Tesseract Planning repository into your workspace: ::
+
+  git clone https://github.com/ros-industrial-consortium/tesseract_planning
+
+If you are using the Robot Operating System (ROS), you can also clone the Tesseract ROS repository: ::
+
+  git clone https://github.com/ros-industrial-consortium/tesseract_ros
+
+Clone the repositories in the dependencies.rosinstall file using wstool or some other method (e.g. manually git cloning them): ::
+
+  wstool init ~/tesseract_ws/src/ ~/tesseract_ws/src/tesseract_planning/dependencies.rosinstall
```
Are these tutorials meant to extend to repositories beyond the core `tesseract` repository? It's not actually necessary to build `tesseract_planning` and `tesseract_ros` in order to build `tesseract`, so we might consider dropping these instructions and putting tutorials for those projects in their respective repositories

marip8 commented on gh_pages/_source/getting_started_doc.rst:
```diff
@@ -0,0 +1,67 @@
+Getting Started
+===============
+
+Install ROS and Catkin
```
`tesseract` is ROS-independent, so the implication that ROS is required to get started with `tesseract` is confusing and  a bit misleading. I would consider revising this section to clarify that ROS and `catkin` make it convenient to build `tesseract`

marip8 commented on gh_pages/_source/getting_started_doc.rst:
```diff
@@ -0,0 +1,67 @@
+
+  source ~/tesseract_ws/devel/setup.bash
+
+NOTE: For noetic tesseract_ext is not required. Install the following dependencies: libbullet-dev, libbullet-extras-dev and ros-noetic-fcl. Taskflow can be install using the PPA below.
```
It's not ROS Noetic that makes `tesseract_ext` not required, it's the fact that the packages in `tesseract_ext` are distributed as binaries on Ubuntu 20.04
```suggestion
> `tesseract_ext` is not required on Ubuntu 20.04 because binary distributions of those packages are available. `taskflow` is the only remaining binary-distributed dependency that must be installed manually. This package can be installed using the PPA below.
```

steviedale replied:
I decided to remove this note and add Taskflow as a step in the installation instructions. I also decided to add a note at the top indicating that the targeted OS for the instructions is Ubuntu 20.04. We may want to create another page/tab/section for installation on 18.04

marip8 commented on gh_pages/_source/tesseract_interface_roscpp_doc.rst:
```diff
@@ -0,0 +1,107 @@
+Tesseract Interface: ROSCPP
```
It seems like this tutorial really belongs in `tesseract_ros`, not in this repository

marip8 commented on gh_pages/_source/tesseract_motion_planning_api_doc.rst:
```diff
@@ -0,0 +1,168 @@
+*****************************
+Tesseract Motion Planning API
```
This tutorial doesn't actually cover the motion planning c++ API as stated; this is more of a tutorial of how to set up a motion planning problem. I think this also belongs either in `tesseract_ros` or `tesseract_planning`, not in this repo

marip8 commented on gh_pages/_source/tesseract_setup_wizard_doc.rst:
```diff
@@ -0,0 +1,275 @@
+**********************
+Tesseract Setup Wizard
```
Same comment; this probably belongs in `tesseract_ignition` rather than this repo

Levi-Armstrong commented:
I believe the plan is to keep everything together but we will have different section for the different repositories. This way there is one place to go for documentation opposed to individual ones for each repository. We also talked about moving this to its own repository which is what MoveIt does.

marip8 commented:
> I believe the plan is to keep everything together but we will have different section for the different repositories. This way there is one place to go for documentation opposed to individual ones for each repository. We also talked about moving this to its own repository which is what MoveIt does.

I think it would make sense for this to be in a separate repository since it's going to have crossover between several other repositories. It would also make it easier to maintain and update since you wouldn't have to run the entire CI pipeline for this repository.

I'm not sure if you've considered this, but it might be worth creating a separate Github organization for `tesseract` for more continuity between all these different repositories

steviedale commented on gh_pages/_source/urdf_and_srdf_doc.rst:
```diff
@@ -0,0 +1,90 @@
+SRDF
+----
+
+The SRDF or Semantic Robot Description Format complement the URDF and specifies joint groups, default robot configurations, additional collision checking information, and additional transforms that may be needed to completely specify the robot's pose. The recommended way to generate a SRDF is using the MoveIt! Setup Assistant.
```
Definitely. Addressed in 98f244f

steviedale commented on gh_pages/_source/urdf_and_srdf_doc.rst:
```diff
@@ -0,0 +1,90 @@
+
+Groups
+^^^^^^
+a 'Group' (sometimes called 'JointGroup' or 'Planning Group') is a central concept in Tesseract. Tesseract always acts on a particular group. Tesseract will only consider moving the joints in the group that it is planning for -- other joints are left stationary. (A motion plan where all joints in the robot may move can be achieved by creating a group containing all joints.) A group is simply a co ...
```
I updated the description in cd5b8ce. How does that sound?

steviedale replied:
I reworded the description for groups in 61858c8. How does that sound?

steviedale commented:
@marip8 and @mpowelson: @KevinWarburton and I really appreciate your feedback on the documentation. If you guys have a little bit of time over the next week or two, it would be great to have you guys continue looking over our changes and supplying more valuable insights. Thanks!

jf--- commented:
>  This way there is one place to go for documentation opposed to individual ones for each repository

This PR is incredible valuable, an amazing effort @steviedale, massive...
Is the above stalling this PR to be merged @Levi-Armstrong ?

mpowelson commented:
@Levi-Armstrong I think I agree. This is certainly better than what we currently have. Could we go ahead and get this merged in the meantime?

steviedale commented:
Thanks @jf---, we're glad to see interest from you and other developers about the Tesseract documentation and excited to get it out there in the wild!

We've decided to put the Sphinx documentation for Tesseract in it's own repository at https://github.com/ros-industrial-consortium/tesseract_docs. We will also separate the documentation into sections: Intro, Core, ROS, ROS2. We're hoping to have the latest changes merged in the next day or two.

steviedale commented:
@Levi-Armstrong I removed the `gh_pages` folder since we've migrated its contents to the `tesseract_docs` repo. I also updated the `README.md` and removed the `sphinx-build` instructions and replaced them with a link to the `tesseract_docs` repo.

The contents of this MR are now:
- removing `gh_pages`
- updating `README.md` with link to `tesseract_docs`
- adding comments with documentation tags to example source code

Ready for review.
